Port hlsdl to SDL3

nspitko opened the pull request:
This is a fairly large/spooky change, but SDL3 offers a lot of features that are extremely useful for modern cross platform development, while also making their existing APIs a lot less gross.

For context: https://wiki.libsdl.org/SDL3/NewFeatures

This has been reworked to be API compatible. The only exception is a change in how controllers are indexed, which will require a code change in heaps to make controllers reliably detect (They still work, but may not pick up without a hot plug in some situations)

heaps PR: https://github.com/HeapsIO/heaps/pull/1302

I'm able to run my jam games without issue, including the multi-window imgui tooling.

Overall the porting cost is extremely low due to how well we're able to hide changes in the hdll. I believe what cost we are going to incur here is well worth the benefits SDL3 offers, especially as we start implementing the new APIs.

I'm opening this as a draft to get more feedback and testing on this. I've tested most of the features I have code coverage for in my own works, but I'd like to get some other heaps users to sign off that their stuff isn't exploding either.

nspitko committed:
Port hlsdl to SDL3

First pass.
Tested major functionality on linux/wayland, seems good.

nspitko committed:
Enable -DDOWNLOAD-DEPENDENCIES for linux

Ubuntu LTS doesn't have libsdl3-dev yet.

nspitko committed:
Remove syswm.h include, add linux deps

nspitko committed:
Remove extra terminator

nspitko committed:
Remove syntactically significant whitespace

nspitko committed:
Remove dead code for HL_MOBILE

nspitko committed:
Update Makefile

nspitko committed:
Try building SDL locally on linux runners

nspitko committed:
Remove setup SDL action, it's broken

Apprentice-Alchemist commented on libs/sdl/sdl.c:
```diff
@@ -746,33 +760,38 @@ DEFINE_PRIM(_I32, win_get_id, TWIN);
 HL_PRIM int HL_NAME(gctrl_count)() {
-	return SDL_NumJoysticks();
+	int count;
+	SDL_GetJoysticks(&count);
```
Calling `SDL_GetJoysticks` without freeing the result leaks memory.

Apprentice-Alchemist commented on libs/sdl/sdl.c:
```diff
@@ -904,66 +928,71 @@ HL_PRIM char* HL_NAME(get_clipboard_text)() {
-	int n = SDL_GetNumVideoDisplays();
-	if (n < 0)
+	int n;
+	SDL_DisplayID *displays = SDL_GetDisplays(&n);
```
Calling `SDL_GetDisplays` without freeing the result leaks memory.

Apprentice-Alchemist commented on libs/sdl/sdl.c:
```diff
@@ -398,34 +399,39 @@ HL_PRIM void HL_NAME(delay)( int time ) {
-	return e.w;
+HL_PRIM int HL_NAME(get_screen_width)( int index ) {
+	int count;
+	SDL_DisplayID *displays = SDL_GetDisplays( &count );
```
Calling `SDL_GetDisplays` without freeing the result leaks memory.

Apprentice-Alchemist commented on libs/sdl/sdl.c:
```diff
@@ -398,34 +399,39 @@ HL_PRIM void HL_NAME(delay)( int time ) {
-	return e.h;
+HL_PRIM int HL_NAME(get_screen_height)( int index ) {
+	int count;
+	SDL_DisplayID *displays = SDL_GetDisplays( &count );
```
Calling `SDL_GetDisplays` without freeing the result leaks memory.

Apprentice-Alchemist commented on libs/sdl/sdl.c:
```diff
@@ -746,33 +760,38 @@ DEFINE_PRIM(_I32, win_get_id, TWIN);
-	return NULL;
+HL_PRIM SDL_Gamepad *HL_NAME(gctrl_open)(int idx) {
+	int count;
+	SDL_JoystickID *sticks = SDL_GetJoysticks(&count);
```
Calling `SDL_GetJoysticks` without freeing the result leaks memory.

Apprentice-Alchemist commented on libs/sdl/sdl.c:
```diff
@@ -808,35 +827,37 @@ DEFINE_PRIM(_I32, haptic_rumble_play, THAPTIC _F64 _I32);
 HL_PRIM int HL_NAME(joy_count)() {
-	return SDL_NumJoysticks();
+	int count;
+	SDL_GetJoysticks(&count);
```
Calling `SDL_GetJoysticks` without freeing the result leaks memory.

Apprentice-Alchemist commented on libs/sdl/sdl.c:
```diff
@@ -904,66 +928,71 @@ HL_PRIM char* HL_NAME(get_clipboard_text)() {
+
+	SDL_DisplayMode **modes;
+	int count;
+	modes = SDL_GetFullscreenDisplayModes(display_id, &count);
```
Calling `SDL_GetFullscreenDisplayModes` without freeing the result leaks memory.

nspitko committed:
Fix memory leaks, manually install SDL on linux

nspitko committed:
Fix build errors

nspitko committed:
SDL_Free -> SDL_free, fix warning

nspitko committed:
Workflow fix

nspitko committed:
SDL_GetJoysticks returns SDL_JoystickID*, not SDL_Joystick*

caught by gcc14 build

nspitko committed:
Remove -I $(BREW_SDL_PREFIX)/include/SDL3, fix IOS

SDL_syswm.h was removed, builders didn't catch because we don't hace CI for IOS/TVOS

nspitko committed:
Fix controller hotplug

SDL3 no longer uses controller index, but a joystick ID now. These IDs are not necessarily sequential, so controllers would often not work unless they were plugged in at launch, or may not work at all if the ids are higher than the count.

This requires a small heaps change as well.

nspitko committed:
Merge branch 'HaxeFoundation:master' into sdl3

nspitko commented:
Honestly I'd rather avoid that. I understand the desire to avoid breaking changes, but I've hidden most of the pain, the stuff that's been changed is largely either a significant alteration to the API that can't be easily hidden (The controller changes), or is actively fixing problems that were affecting me already (SDL window related functions not taking window pointers, making multi-window apps problematic)

I can probably reduce a break or two here and there with defaults, but I don't think this is going to be worth the time to merge if we're not willing to accept some level of API breakage, and I don't think there's any world in which a total rewrite of hlsdl makes sense right now, given the work it would put on heaps.

I fully agree that the sdl implementation in this lib is a bit of a mess. That said, right now it basically only exists to enable heaps and libs based off heaps, so unless we're going to get sign-on from the heaps side to massively rework the SDL backend, I don't see that going anywhere. The PR on the heaps is very small and heaps users already deal with API breakages regularly between sdl and fmt.

nspitko committed:
Maintain API compatibility with SDL2

nspitko committed:
Merge branch 'sdl3' of https://github.com/nspitko/hashlink into sdl3

nspitko committed:
Add missing window flags, fix SDL_SetEventEnabled

nspitko committed:
Emulate SDL_GRAB_KEYBOARD

nspitko committed:
Add missing cast

nspitko committed:
Fix missing bool

nspitko commented:
Reworked things to not break API. I'm not going to go as far as to ensure no behavior has changed; doing this "right" would mean pulling in thousands of lines from sdl_compat.c and that's just way outside of the scope.

From my testing now everything works as it did before in heaps, cerastes, and imgui without code changes. The one exception is controllers require a fix in heaps due to the ID changes to detect reliably on all platforms. Fixing this requires forcing hlsdl to internally manage controllers so it can hold a stable index, which is beyond the scope of what I'm willing to do here. We should just make the code change in heaps.

this does NOT add any new functions (outside of SDL_Error which is/was useful for debugging issues that crop up during porting), those will be done in a separate PR to limit the scope of this change as much as possible.

nspitko commented:
> Not sure I fully understand the details or why we'd need to manage anything. Can't we use SDL_GetJoysticks to map from index to id? Like this: https://github.com/tobil4sk/hashlink/commit/ffa1f61fdabbce43563248910c9059935e4d8852

Because the return order of SDL_GetJoysticks is not stable across all platforms. In my testing it very specifically was not stable and did not work properly. 

I'm explicitly not adding new SDL3 functions to the haxe API in this PR. The heaps change fixes the bug without breaking API, and is compatible with both SDL2 and SDL3. This only even exists because heaps doesn't initialize controllers during launch, so if there are any other consumers using this API they likely aren't affected.

I totally understand the desire to sdl3-ify stuff and clean up the API but that's out of scope for this change.

nspitko committed:
jdevice->gdevice

This doesn't affect anything as these point to the same field, but will prevent a bug in the future if for some reason these structs become unaligned in future SDL releases.

nspitko commented on libs/sdl/sdl.c:
```diff
@@ -239,141 +293,93 @@ HL_PRIM bool HL_NAME(event_loop)( event_data *event ) {
-		case SDL_CONTROLLERDEVICEREMOVED:
+		case SDL_EVENT_GAMEPAD_REMOVED:
 			event->type = GControllerRemoved;
 			event->reference = e.jdevice.which;
```
Fixed. This doesn't actually affect anything since they point to the same field, but would fix a bug if someone in the future changed the struct alignment. (I don't think this would ever happen since these are very intentionally aligned with eachother, but better safe than sorry)

nspitko commented:
> In that case can we just store the joystick array statically after the gctrl_count call so it is then stable for subsequent calls to gctrl_open? This seems to be sufficient for sdl2-compat, so I think it should be a fairly reasonable solution.

This is fragile and relies on very specific calling conventions. heaps has multiple paths for initializing joysticks depending on whether it's at launch or at runtime, the latter of which never calls gctrl_count. You'd need to internally build and track them during the SDL connect/disconnect events, then have gctrl_count return the arraysize here; but this still doesn't solve for cases where the controller count changes between event dispatch and the gctrl_open call. Some composite devices do weird stuff like this; I imagine this is why the API was changed in the first place.

> Understood, however, I'm not sure there is much point in making heaps compatible with this in-between state where the sdl2 count-based iteration has been broken, but the sdl3 api to replace it hasn't been exposed yet.

Seems fine to me? I don't expect anyone is going to be in a hurry to rewrite the heaps controller code even if we expose the SDL3 controller API immediately. It works fine, minus this one bug we're about to create. The fix is simple and doesn't break sdl2 compat.

nspitko committed:
add getJoysticks. Bump version.

nspitko committed:
SDL_Joystick -> SDL_JoystickID

nspitko committed:
Update haxelib.json

nspitko committed:
fix text input

In SDL2, windows start with text_input enabled. SDL3 toggles this, so we need to enable text input whenever we create a window to avoid breaking existing apps.

nspitko committed:
Merge remote-tracking branch 'upstream/master' into sdl3

yuxiaomao committed:
Merge branch 'master' into sdl3

yuxiaomao committed:
Fix spaces, format, remaining sdl2

yuxiaomao committed:
Make getJoysticks optional, use displayId as handle for easier migration

yuxiaomao committed:
Bump SDL to 3.4.10 and fix copilot setup

yuxiaomao committed:
Fix fullscreen, remove dup code

yuxiaomao committed:
Add missing nullcheck and free, do not expose NativeArray

yuxiaomao committed:
Adjust linebreaks
